File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys

sys.setrecursionlimit(10000)



def dfs(ox,oy) : 
  cx = [1, -1, 0, 0]
  cy = [0, 0, 1, -1]

  for i in range(4):
    X = ox + cx[i]
    Y = oy + cy[i]

    if (0 <= X < N) and (0 <= Y < M) : 
      if graph[X][Y] == 1 : 
        graph[X][Y] = 2 
        dfs(X,Y)
    

total = int(input())
for _ in range(total) : 
  M,N,K = map(int, input().split())
  
  cnt = 0
  graph = [[0] * (M) for _ in range(N)]

  for _ in range(K) : 
    y,x = map(int, input().split())
    graph[x][y] = 1 

  for i in range(N):
    for j in range(M): 
      if graph[i][j] == 1 : 
        dfs(i,j)
        cnt += 1
  print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const total = Number(input.shift());

let board;
let m, n, k;
let answer = [];

for (let t = 0; t < total; t++) {
  let cnt = 0;
  [m, n, k] = input.shift().split(" ").map(Number);

  board = [...new Array(n)].map((_) => new Array(m).fill(0));

  for (let i = 0; i < k; i++) {
    //   가로(y)  세로(x)
    let [mNode, nNode] = input.shift().split(" ").map(Number);
    board[nNode][mNode] = 1;
  }

  for (let i = 0; i < n; i++) {
    // 8
    for (let j = 0; j < m; j++) {
      // 10
      if (board[i][j] == 1) {
        dfs(i, j);
        cnt++;
      }
    }
  }

  answer.push(cnt);
}

for (let i = 0; i < answer.length; i++) {
  console.log(answer[i]);
}

function dfs(x, y) {
  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  board[x][y] = 0;

  for (let i = 0; i < 4; i++) {
    let X = x + dx[i];
    let Y = y + dy[i];

    if (X >= 0 && X < n && Y >= 0 && Y < m && board[X][Y] !== 0) {
      dfs(X, Y);
    }
  }
}
